### PR TITLE
FIR/UAST: commonize Identifier implementations

### PR DESCRIPTION
**Pull request by jsjeon**
This is a continued work to commonize declarations that are (close to or) in final shape. #4362 did for `UFile` implementations; #4384 did for `UImportStatement` implementations; and this PR is aiming for identifiers implementations: `UastLightIdentifier` and `KotlinUIdentifier`.

1st commit first utilizes FE1.0 UAST's identifer test base so that FIR UAST can share the utils. Those utils are migrated to the base module's tests.

2nd commit introduces identifiers and reference names tests to see the current state of FIR UAST in terms of, well, literally identifiers and reference names. It also comprehensively tests legacy UAST against all of legacy test data. A good news is almost all tests work properly; and a bad news is that there are three cases where an expected identifier conversion failed: see the whitelist in `AbstractFE1LegacyUastIdentifiersTest`.

Ironically, the state of FIR UAST's identifier conversion was... empty. :\ With the test preparation, I found that `LeafPsiElement` to `UIdentifier` conversion was completely missed. 3rd commit adds that, and now we can see some results. As expected, several parts are missing: expression, annotations, etc.

Then, 4th commit commonizes identifiers implementations. They don't need sharing of super-type hierarchy (like `UElement`/`UExpression` do), so relatively trivial.

### CODE REVIEW

**Comment by jsjeon** [comment]
Will keep this as a draft until #4384 is merged, since commonized `UnknownKotlinExpression` will affect many test outputs. Also, as test metadata for identifers/refNames are separated in `testData/legacyIdentifiers`, I'd like to rename existing `testData/legacy` to `testData/legacyRenderLog`, which also needs #4384 being merged first.

File: plugins/uast-kotlin-fir/src/org/jetbrains/uast/kotlin/FirKotlinConverter.kt [hunk 1]
```diff
@@ -192,6 +195,25 @@ internal object FirKotlinConverter {
                 is KtImportDirective -> {
                     el<UImportStatement>(build(::FirKotlinUImportStatement))
                 }
+                is LeafPsiElement -> {
+                    when {
+                        element.elementType in identifiersTokens -> {
+                            if (element.elementType != KtTokens.OBJECT_KEYWORD ||
```
Comment (tgeng) [inline plugins/uast-kotlin-fir/src/org/jetbrains/uast/kotlin/FirKotlinConverter.kt:201]: Can you invert this if condition? Also, shouldn't the second condition be `!= null`?

Comment (jsjeon) [reply plugins/uast-kotlin-fir/src/org/jetbrains/uast/kotlin/FirKotlinConverter.kt:201]: Re: inverting the condition,

This came from the counterpart in FE1.0 UAST, so I prefer to stick to the original form (or to change both sides together).

Re: the 2nd `null` check condition,

IIUC, these lines are read: build `UIdentifier` if the `element` isn't (reserved) `object` keyword, or even if it's `object` keyword, but not from `object` declaration (e.g., escaped with back ticks?).

Anyway, switching to `!= null` shows that all of `object` appeared in the tests are popped as identifers.

**Comment by jsjeon** [comment]
This PR is ready: cc @darthorimar @xiexed

**Comment by jsjeon** [comment]
> Also, as test metadata for identifers/refNames are separated in `testData/legacyIdentifiers`, I'd like to rename existing `testData/legacy` to `testData/legacyRenderLog`, which also needs #4384 being merged first.

Done as the 6th commit. 5th commit is placing each `Identifier` in separate files.

File: plugins/uast-kotlin-fir/tests/org/jetbrains/uast/test/kotlin/AbstractFirLegacyUastIdentifiersTest.kt [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+    override fun isExpectedToFail(filePath: String): Boolean {
+        // TODO: investigate why everything is failing!
+        return true
+    }
+
+    override fun check(filePath: String, file: UFile) {
+        if (filePath == "plugins/uast-kotlin/testData/EnumValueMembers.kt") {
+            // TODO: need a special handling of members in enum entries. See testEnumValueMembers in FirLegacyUastIdentifiersTestGenerated
+            //   Seems related to KT-45115: in FIR, enum entry init is an anonymous object, and thus members in it are regarded as _local_.
+            //   whereas FIR UAST conversion took a path to look up non-local FIR declaration for now.
```
Comment (jsjeon) [inline plugins/uast-kotlin-fir/tests/org/jetbrains/uast/test/kotlin/AbstractFirLegacyUastIdentifiersTest.kt:21]: cc @mglukhikh @darthorimar  I'll take a look at this afterwards.

**Review by darthorimar — APPROVED** [review]

**Review by xiexed — APPROVED** [review]
Except the fact that I hate test generation as an approach everything else looks good to me.

**Comment by jsjeon** [comment]
> Except the fact that I hate test generation as an approach everything else looks good to me.

Oops, sorry to hear that. I assume you don't like text-based file comparison, and yeah it's adding too much noises. But, as my self defense, at the current stage, I want to make sure the refactoring doesn't break FE1.0 UAST. Although `:plugins:uast-kotlin` has a number of tests, I'm a bit concerned about its coverage, and thus added a bunch of tests as a base line. Once settled down, we could remove these additions, and instead add API tests. Also, one good thing about the current infra is, it's relatively easy to add new tests, e.g., if I found issues while integrating FIR UAST with Android Lint or metalava.

Actually... for further refactoring (like commonizing some trivial expression conversions as well as type-related expressions and their type conversions), I'm adding more and more tests like these, so I'm worrying if you don't like those either. :\ Any suggestions, opinions, pointers, etc. are welcome.

**Comment by xiexed** [comment]
It is not a big deal, please continue as is most convenient for you. Just before it there were no generated tests in UAST and I  liked that fact, but it is just a personal taste not a strong position. Anyway we will be able to refactor it at later stages

**Comment by jsjeon** [comment]
FYI, last commit was rebased on top of https://github.com/JetBrains/kotlin/commit/65972fa64ce10b659480d1a6fb8c3470205d5c62

**Pull request merged**